src/dodal/devices/oav/oav_detector.py

```diff
@@ -30,63 +39,89 @@ class ZoomController(Device):
-    onst = Component(EpicsSignal, "MP:SELECT.ONST")
-    twst = Component(EpicsSignal, "MP:SELECT.TWST")
-    thst = Component(EpicsSignal, "MP:SELECT.THST")
-    frst = Component(EpicsSignal, "MP:SELECT.FRST")
-    fvst = Component(EpicsSignal, "MP:SELECT.FVST")
-    sxst = Component(EpicsSignal, "MP:SELECT.SXST")
-
-    @property
-    def allowed_zoom_levels(self):
-        return [
-            self.zrst.get(),
-            self.onst.get(),
-            self.twst.get(),
-            self.thst.get(),
-            self.frst.get(),
-            self.fvst.get(),
-            self.sxst.get(),
-        ]
-
-    def set(self, level_to_set: str) -> StatusBase:
-        return self.level.set(level_to_set)
-
-
-class OAV(AreaDetector):
-    cam = ADC(CamBase, "-DI-OAV-01:CAM:")
-    roi = ADC(ROIPlugin, "-DI-OAV-01:ROI:")
-    proc = ADC(ProcessPlugin, "-DI-OAV-01:PROC:")
-    over = ADC(OverlayPlugin, "-DI-OAV-01:OVER:")
-    tiff = ADC(OverlayPlugin, "-DI-OAV-01:TIFF:")
-    hdf5 = ADC(HDF5Plugin, "-DI-OAV-01:HDF5:")
-    grid_snapshot = Component(SnapshotWithGrid, "-DI-OAV-01:MJPG:")
-    snapshot = Component(SnapshotWithBeamCentre, "-DI-OAV-01:MJPG:")
-    zoom_controller = Component(ZoomController, "-EA-OAV-01:FZOOM:")
-
-    def __init__(self, *args, params: OAVConfigParams, **kwargs):
-        super().__init__(*args, **kwargs)
-        self.parameters = params
-        self.grid_snapshot.oav_params = params
-        self.snapshot.oav_params = params
-        self.subscription_id = None
-        self._snapshot_trigger_subscription_id = None
-
-    def wait_for_connection(self, all_signals=False, timeout=2):
-        connected = super().wait_for_connection(all_signals, timeout)
-        x = self.grid_snapshot.x_size.get()
-        y = self.grid_snapshot.y_size.get()
-
-        cb = partial(self.parameters.update_on_zoom, xsize=x, ysize=y)
-
-        if self.subscription_id is not None:
-            self.zoom_controller.level.unsubscribe(self.subscription_id)
-        self.subscription_id = self.zoom_controller.level.subscribe(cb)
-
-        return connected
+    def __init__(self, prefix: str, name: str = "") -> None:
+        super().__init__(name=name)
+        self.percentage = epics_signal_rw(float, f"{prefix}ZOOMPOSCMD")
+
+        # Level is the string description of the zoom level e.g. "1.0x" or "1.0"
+        self.level = epics_signal_rw(str, f"{prefix}MP:SELECT")
+
+    async def _get_allowed_zoom_levels(self) -> list:
+        zoom_levels = await self.level.describe()
+        return zoom_levels["level"]["choices"]  # type: ignore
+
+    @AsyncStatus.wrap
+    async def set(self, level_to_set: str):
+        allowed_zoom_levels = await self._get_allowed_zoom_levels()
+        if level_to_set not in allowed_zoom_levels:
+            raise ZoomLevelNotFoundError(
+                f"{level_to_set} not found, expected one of {allowed_zoom_levels}"
+            )
+        await self.level.set(level_to_set, wait=True)
```

Comment:
I think long term we shouldn't be doing this check. I've added https://github.com/DiamondLightSource/dodal/issues/884